apps/boxclk/app.js

```diff
@@ -0,0 +1,223 @@
+  const month = locale.month(date, 1);
+  const year = date.getFullYear();
+  let suffix;
+  if ([1, 21, 31].includes(dayOfMonth)) {
+    suffix = "st";
+  } else if ([2, 22].includes(dayOfMonth)) {
+    suffix = "nd";
+  } else if ([3, 23].includes(dayOfMonth)) {
+    suffix = "rd";
+  } else {
+    suffix = "th";
+  }
+  let dayOfMonthStr = enableSuffix ? dayOfMonth + suffix : dayOfMonth;
+  return month + " " + dayOfMonthStr + ", " + year;
+}
+
+function getDayOfWeek(date) {
+  return locale.dow(date, 0);
+}
+
+function draw(boxes) {
+  date = new Date();
+  g.clear();
+  if (bgImage) {
+    g.drawImage(bgImage, 0, 0);
+  }
+  if (boxes.time) {
+    boxes.time.string = locale.time(date, 1);
+  }
+  if (boxes.date) {
+    boxes.date.string = getDate();
+  }
+  if (boxes.dow) {
+    boxes.dow.string = getDayOfWeek(date);
+  }
+  if (boxes.batt) {
+    boxes.batt.string = E.getBattery() + "%";
+  }
+  Object.keys(boxes).forEach((boxKey) => {
+    let boxItem = boxes[boxKey];
+    calcBoxSize(boxItem);
+    const pos = calcBoxPos(boxKey);
+    if (isDragging[boxKey]) {
+      g.setColor(boxItem.border);
+      g.drawRect(pos.x1, pos.y1, pos.x2, pos.y2);
+    }
+    g.drawString(
+      boxItem,
+      boxItem.string,
+      boxPos[boxKey].x +  boxItem.xOffset,
+      boxPos[boxKey].y +  boxItem.yOffset
+    );
+  });
+  if (!Object.values(isDragging).some(Boolean)) {
+    if (drawTimeout) clearTimeout(drawTimeout);
+    drawTimeout = setTimeout(() => draw(boxes), 60000 - (Date.now() % 60000));
+  }
+}
+
+function setup() {
+  // Define the touchHandler function
+  touchHandler = function(zone, e) {
+    wasDragging = Object.assign({}, isDragging);
+    let boxTouched = false;
+    Object.keys(boxes).forEach((boxKey) => {
+      if (touchInText(e, boxes[boxKey], boxKey)) {
+        isDragging[boxKey] = true;
+        wasDragging[boxKey] = true;
+        boxTouched = true;
+      }
+    });
+    if (!boxTouched) {
+      Object.keys(isDragging).forEach((boxKey) => {
+        isDragging[boxKey] = false;
+      });
+    }
+    if (Object.values(wasDragging).some(Boolean) || !boxTouched) {
+      draw(boxes);
+    }
+  };
+
+  // Define the dragHandler function
+  dragHandler = function(e) {
+    Object.keys(boxes).forEach((boxKey) => {
+      if (isDragging[boxKey]) {
+        let boxItem = boxes[boxKey];
+        calcBoxSize(boxItem);
+        let newX = boxPos[boxKey].x + e.dx;
+        let newY = boxPos[boxKey].y + e.dy;
+        if (newX - totalWidth / 2 >= 0 &&
+            newX + totalWidth / 2 <= w &&
+            newY - totalHeight / 2 >= 0 &&
+            newY + totalHeight / 2 <= h ) {
+          boxPos[boxKey].x = newX;
+          boxPos[boxKey].y = newY;
+        }
+        const pos = (boxKey);
+        g.clearRect(pos.x1, pos.y1, pos.x2, pos.y2);
+      }
+    });
+    draw(boxes);
+  };
+
+  Bangle.on('touch', touchHandler);
+  Bangle.on('drag', dragHandler);
+
+  Bangle.setUI({
+    mode : "clock",
+    remove : function() {
```

Comment:
Nice clock! Just thinking, I believe for the fastloading you'd also want to put the variables of this file inside a scope, so they're removed when the clock's unloaded (or technically once the code's run and any closures over them are gc'd)

Reply:
Yes - check out http://www.espruino.com/Bangle.js+Fast+Load for more info - all you need to do is put a `{}` around everything and then define your global functions as `let functionName = function() { ... }`

Reply:
Thanks! It passes the memory leak test after wrapping the entire program in an IIFE. I think this is sufficient, right?

```
>Bangle.loadWidgets();Bangle.drawWidgets();
=undefined
=2242
=undefined
>process.memory().usage
=3549
=undefined
=2242
```

Reply:
Yeah, it's ok - but because of the way Espruino works it ends up parsing the same code twice - once to load the function and once to run it, so it's quite a bit slower to load than my suggestion.

But if you're fine with this I'm ok to merge it